### PR TITLE
add Setup Test Page button into test window

### PR DESCRIPTION
**Pull request by mzgoddard**
[Open PR Preview](https://raw.githack.com/mzgoddard/aria-at/mzgoddard-reset-test/index.html)

This change injects a button at the bottom of the test page in the window opened by activating the `Open Test Page` button.

The button has the text `Setup Test Page` and the autofocus attribute. Clicking it will run a test setup script if this test has one and then disables itself. If there is no setup script the button will disabled at the start.

<details>
<summary>Original PR Description</summary>

This change is a first step for https://github.com/w3c/aria-at/issues/358. In a follow up PR, we can inject a button into the test page that on click will reload the test page window.

After the test page window is reloaded, or refreshed, the setupTestPage script will be executed on the reloaded page. As well the unload listener is added again so that a second reload, and third and so on, will have the same functionality. By adding the unload listener after reloading it will also be able to re-enable the open test page when it is closed.
</details>

### CODE REVIEW

**Comment by jscholes** [comment]
@mzgoddard Sorry, but this isn't the approach the community group reached consensus on.  #358 is outdated, and #369 is the more appropriate reference.  In summary:

1. Setup scripts should never be executed on page load.  This causes screen reader issues, script flakiness, among other things.
2. There should be a button on the test page that runs the setup script, which testers will need to manually press.
3. Once pressed, the button can transition to being a "Reset" button, which only refreshes the test page and allows the execute button to be pressed again.  If a tester chooses to manually refresh the page instead to restore the execute button, that's also fine.

**Comment by s3ththompson** [comment]
@jscholes @mzgoddard I think Z plans to add the actual button (with the setup / reset behavior) in a followup PR. We discussed the consensus from #369 and again as summarized in your issue. I think perhaps the point of confusion was: how can we prevent the script that dynamically injects the setup script button from being flaky itself? Do you have any context on how to safely inject a new button element into the example?

**Comment by jscholes** [comment]
@s3ththompson Unless the wording of the initial comment on this PR is inaccurate, I don't agree with your assessment.  Specifically:

> After the test page window is reloaded, or refreshed, the setupTestPage script will be executed on the reloaded page.

... and ...

> In a follow up PR, we can inject a button into the test page that on click will reload the test page window.

What I'm reading here is:

1. this PR allows the setup script to be executed on refresh, and
2. a follow-up PR will add an explicit refresh button, which relies on the behaviour introduced in this one, and
3. testers will only need to press that explicit button once, because it will reload the page and exec the setup script in one go.

If my understanding is wrong, please let me know and update the PR description.  Otherwise, if I'm reading the situation correctly, I maintain that this isn't the approach we want to proceed with.

File: tests/resources/aria-at-harness.mjs [hunk 1]
```diff
@@ -161,10 +162,38 @@ function executeScriptInTestPage() {
...
+  const buttonDiv = testPageWindow.document.createElement('div');
+  buttonDiv.innerHTML = `
+    <div style="position: relative; left: 0; right: 0; height: 2rem;">
+      <button style="height: 100%; width: 100%;">Run Test Setup</button>
+    </div>
+  `;
+  const button = buttonDiv.querySelector('button');
+
+  /** @type {'setup' | 'reload'} */
+  let runSetupOrRelad = 'setup';
```
Comment (jscholes) [inline tests/resources/aria-at-harness.mjs:174]: You have a small typo here, "Relad" -> "Reload".

**Comment by jscholes** [comment]
@mzgoddard I notice that there is no preview for this PR.  What's going on there?

**Comment by mzgoddard** [comment]
> @mzgoddard I notice that there is no preview for this PR. What's going on there?

I think that was this PR not having rebased onto the #441 revert (by #449). Unless you are referring to something else.

**Comment by jscholes** [comment]
That definitely sounds like the likely cause.  Just noting that previewing this functionality will be critical.

**Comment by mzgoddard** [comment]
@jscholes I updated the PR's description. I updated this change to inject a button to the top of the test page that on a first click runs the setup script if there is one, and on a second click reloads the page.

I suspect this still has some of the execution flakkiness of the current setup script execution. I've seen in my testing on Firefox, on Mac, that on test page reload the button may not inject. I suspect this has to do with the way the setup script was automatically run and now how the button is injected. I haven't seen this in Chrome, but that may be due to a smaller sample size as I'm mostly testing in Firefox.

And I think it may cause some issue with screen readers with the button being injected.

If these issues do persist in this change we probably have to change how the button is inserted, primarily when it is inserted. The easiest option may be to modify the reference pages that tests are performed in, but I am trying to avoid that.

**Comment by jscholes** [comment]
@mzgoddard Maybe I'm missing something?  I looked at some tests from the Modal Dialog plan, both via the Index and Review pages.  But the setup scripts are still automatically executed when I open the test page, and there are no new buttons anywhere.

**Comment by richnoah** [comment]
@jscholes would you please convey what URL you are using to view the tests for the Modal Dialog Plan to aid in diagnosing the issue?

**Comment by jscholes** [comment]
@richnoah The preview page at: https://raw.githack.com/mzgoddard/aria-at/mzgoddard-reset-test/review/modal-dialog.html

**Comment by mzgoddard** [comment]
@jscholes Ah. This should have been obvious to me. I had been confirming the change on test pages and not on review pages.

The review page doesn't use the aria-at-harness module to open the test page window. So it currently needs its own change to the script it uses.

**Comment by jscholes** [comment]
@mzgoddard Makes sense.  Is it planned for the review pages to use the harness to cut down on duplication and other corner cases in the future?

In terms of the implementation of the button when using test pages:

1. The button needs an `autofocus` attribute to ensure that it receives focus automatically at page load.  I stress that it should use `autofocus`, rather than `element.focus`, as it's then up to the browser to ensure that it works correctly.
2. The button isn't within the main landmark region, which will cause extra speech verbosity when activating it moves focus onto something within main.  Can it be injected into the `<main>` container?

Other than the above, it seems to be working great.

**Comment by s3ththompson** [comment]
> The button isn't within the main landmark region, which will cause extra speech verbosity when activating it moves focus onto something within main. Can it be injected into the <main> container?

@jscholes It would be great if this script was agnostic to the format of the example (to prepare for examples beyond APG...). Will all examples have `main` elements? Or should the button injection code try to find a `main` element, but fall back to the `body`?

> I've seen in my testing on Firefox, on Mac, that on test page reload the button may not inject.

@mzgoddard I am able to reproduce this issue consistently on Chrome. The 100ms timeout seems like it might be an issue. Can the parent asynchronously wait until the window has been loaded (or reloaded) before triggering (or retriggering) the script?

**Comment by jscholes** [comment]
> should the button injection code try to find a main element, but fall back to the body?

This sounds like a reasonable approach.

> Can the parent asynchronously wait until the window has been loaded (or reloaded) before triggering (or retriggering) the script?

I'm concerned about the use of `autofocus`, which is important to the intended design, if the button isn't present at page load.  Is it not possible to somehow construct the mark-up before presenting it to be rendered, rather than "injecting" the button into an existing document?

**Comment by mzgoddard** [comment]
Was able to revisit this and made some changes. A major change is removing the attempt to support reloading the window. More on that after a list of other differences.

Changes pushed to this PR:
- add button in `main` element if it exists, otherwise add to`body` element
- add button on DOMContentLoaded event, when the dom can first be modified
- use TestRunWindow in review pages, using the same code to open the window and add setup button
- disable setup button if there is no setup script or once it is activated
- add autofocus attribute to the setup button

I think after visiting this a few times, before we can use a technical solution to support "resetting" the test page we probably need a change to the process of making test plans. I think there are two non-exclusive changes that could be made to the process of making test plans. I'll write about them in #369 as we can treat it as a separate set of work.

**Comment by jscholes** [comment]
@mzgoddard 

> I think after visiting this a few times, before we can use a technical solution to support "resetting" the test page we probably need a change to the process of making test plans.

I'm intrigued, so I'll eagerly await updates on #369.  Right now, it seems like one of the simple parts of the web, i.e. refreshing a page, doesn't work because of how the windows are being opened, kind of as pop-ups rather than straight pages.  Is there really no way we can just have the page open by URL, and include some JS files that can handle setup script duties?  That would also make testing them a lot easier, because it would mean not having to go through the testing flow to get there.
